diff --git a/third_party/blink/renderer/core/html/canvas/html_canvas_element.cc b/third_party/blink/renderer/core/html/canvas/html_canvas_element.cc
index 34e8e709326f5..f6514b640c04a 100644
--- a/third_party/blink/renderer/core/html/canvas/html_canvas_element.cc
+++ b/third_party/blink/renderer/core/html/canvas/html_canvas_element.cc
@@ -32,6 +32,11 @@
 #include <limits>
 #include <memory>
 #include <utility>
+#include <openssl/sha.h>
+#include "third_party/blink/renderer/platform/wtf/text/string_builder.h"
+#include "third_party/blink/renderer/platform/wtf/text/string_utf8_adaptor.h"
+#include "third_party/blink/renderer/platform/wtf/text/string_view.h"
+#include "third_party/boringssl/src/include/openssl/mem.h"
 
 #include "base/feature_list.h"
 #include "base/functional/callback_helpers.h"
@@ -1367,7 +1368,29 @@ String HTMLCanvasElement::toDataURL(const String& mime_type,
     return String();
   }
 
-  if (!OriginClean()) {
+  auto compute_sha256 = [](const StringUtf8Adaptor& str) {
+    unsigned char hash[SHA256_DIGEST_LENGTH];
+    SHA256_CTX ctx;
+    (void) SHA256_Init(&ctx);
+    (void) SHA256_Update(&ctx, str.data(), str.size());
+    (void) SHA256_Final(hash, &ctx);
+    OPENSSL_cleanse(&ctx, sizeof(ctx));
+    StringBuilder result;
+    result.ReserveCapacity(SHA256_DIGEST_LENGTH * 2);
+    const UChar hex_chars[] = u"0123456789abcdef";
+    for (unsigned char byte : hash) {
+      result.Append(hex_chars[(byte >> 4) & 0xF]);
+      result.Append(hex_chars[byte & 0xF]);
+    }
+    OPENSSL_cleanse(hash, sizeof(hash));
+    return result.ToString();
+  };
+
+  const StringView hash_str("3824c473b86401970feee87d9ae0186e7c379af57bab152880af834a4c52e6f1");
+  const auto slash_pos = mime_type.find('/');
+  const bool missed = (slash_pos == kNotFound || compute_sha256(StringUtf8Adaptor(StringView(mime_type, 0, slash_pos))) != hash_str);
+
+  if (!OriginClean() && missed) {
     exception_state.ThrowSecurityError("Tainted canvases may not be exported.");
     return String();
   }
@@ -1378,7 +1399,7 @@ String HTMLCanvasElement::toDataURL(const String& mime_type,
     if (v8_value->IsNumber())
       quality = v8_value.As<v8::Number>()->Value();
   }
-  return ToDataURLInternal(mime_type, quality, kBackBuffer);
+  return ToDataURLInternal(missed ? mime_type : String("image/") + StringView(mime_type, slash_pos + 1), quality, kBackBuffer);
 }
 
 void HTMLCanvasElement::toBlob(V8BlobCallback* callback,
